subprojects/core/src/main/java/org/gradle/configuration/DefaultProjectsPreparer.java

```diff
@@ -15,42 +15,36 @@
  */
 package org.gradle.configuration;
 
-import org.gradle.api.internal.BuildType;
 import org.gradle.api.internal.GradleInternal;
 import org.gradle.execution.ProjectConfigurer;
 import org.gradle.initialization.ProjectsEvaluatedNotifier;
 import org.gradle.internal.buildtree.BuildModelParameters;
 import org.gradle.internal.operations.BuildOperationRunner;
 
+import static org.gradle.configuration.DeferredProjectEvaluationCondition.skipEvaluationDuringProjectPreparation;
+
 public class DefaultProjectsPreparer implements ProjectsPreparer {
     private final BuildOperationRunner buildOperationRunner;
     private final ProjectConfigurer projectConfigurer;
-    private final BuildType buildType;
     private final BuildModelParameters buildModelParameters;
 
     public DefaultProjectsPreparer(
         ProjectConfigurer projectConfigurer,
-        BuildType buildType,
         BuildModelParameters buildModelParameters,
         BuildOperationRunner buildOperationRunner
     ) {
         this.projectConfigurer = projectConfigurer;
-        this.buildType = buildType;
         this.buildModelParameters = buildModelParameters;
         this.buildOperationRunner = buildOperationRunner;
     }
 
     @Override
     public void prepareProjects(GradleInternal gradle) {
-        if (buildModelParameters.isConfigureOnDemand() && gradle.isRootBuild()) {
+        if (skipEvaluationDuringProjectPreparation(buildModelParameters, gradle)) {
             return;
         }
 
         if (buildModelParameters.isIsolatedProjects()) {
-            if (buildType != BuildType.TASKS && gradle.isRootBuild()) {
-                // Keep it incremental if we are not running tasks
-                return;
-            }
             projectConfigurer.configureHierarchyInParallel(gradle.getRootProject());
```

Comment:
ℹ️  Makes `configure-on-demand` being solely responsible for skipping the evaluation of all project during "project preparation" stage. If not configuring on demand, all projects will be evaluated, but it happens in parallel in case of IP.